Rewrite the check for alias simplification in auto_functionalize_v2 in a way that guarantee no guards are created

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138081

The alias simplification is best effort, its better not to introduce guards than to do the alias simplification. 
if the view is originally coming from an alias operation, MOST likely we should be able to statically know that, if Not
Well its ok then not to use alias. 

this way we guarantee that there are no guards added.

laithsakka committed:
rewrite the check for alias simplficiaiton in auto_functionalize_v2 in a way that guarantee no guards are created

[ghstack-poisoned]

laithsakka committed:
Update on "rewrite the check for alias simplficiaiton in auto_functionalize_v2 in a way that guarantee no guards are created"




[ghstack-poisoned]

laithsakka committed:
Update on "Rewrite the check for alias simplification in auto_functionalize_v2 in a way that guarantee no guards are created"


The alias simplification is best effort, its better not to introduce guards than to do the alias simplification. 
this way we guarantee that there are no guards added. 


[ghstack-poisoned]

ezyang commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -84,16 +84,20 @@ def write_single_view(prefix: str, tensor: Tensor, base_index: int):
+                sym_eq(base.stride(), tensor.stride()),
+                sym_eq(base.size(), tensor.size()),
             ):
                 kwargs[f"{prefix}_alias"] = True
```
@bdhirsh do you know if a true guard is needed here? Because if it's OK to miss on the alias here if it's not "obvious" that the base and tensor are exactly the same offset/stride/size, then we don't even need `parallel_and_statically_known_true`

laithsakka replied:
This is actually a code that i added in functionalization V2, I need to write a post explaining how it works but in short here, by default every view is regenerated by calls to as_strided(). 

For two reasons we actually want to generate alias instead of as_strided when possible: 
1. Readability and simplicity. 
2. chatting with @Elias Ellison alias is better optimized by inductor than as_strided since as_strided enforce realization, and other optimizations can potentially reasons easier about the program with alias.

I added a test to make sure we do not do recompilation when i added that originally and we had some recompilations due to  base.size()== tensor.size(), re-ordering the checks prevented that. But this way is more sound. if something not caught by my tests.  

Why is it safe not to guard?
if the path is taken ==> Then we statically know its True, so it will always be True and hence no guards needed.
if the path is NOT taken => we generate as_strided which works anyway always, and in that case the probability that this view is originally is coming from an alias is low, because if so we should have  parallel_and_statically_known_true works.

lmk if I am missing anything or my explanation is not clear.

laithsakka replied:
rewrite to to
```
 all(
                statically_known_true(a)
                for a in [
                    sym_eq(base.storage_offset(), tensor.storage_offset()),
                    sym_eq(base.stride(), tensor.stride()),
                    sym_eq(base.size(), tensor.size()),
                ]
 ```

ezyang replied:
The rewritten version looks good, we do this other places in aot autograd too

bdhirsh replied:
side-car comment, but I'm a bit curious why we need the `as_strided` at all, since functionalization already stores the exact set of views needed to regenerate a given alias from its base

zou3519 replied:
We end up reconstructing the view in Inductor, so the functionalization view info isn't available.

Ideally we'd pass a subgraph that represents how to regenerate the view to the auto_functionalized HOP and use that to regenerate the view. We didn't do this (because it required some more infra); the alternative to reconstructing the view is to store the sizes/strides/storage_offset.

The plan on record is for someone (TBD) to actually do the work of constructing a subgraph of the view when they run into one of our limitations:
1) complicated views where we can't infer that it's an alias/slice/select
2) as_strided causing potential issues (like inductor needing to realize for exact strides)
3) Whenever export decides they want to use auto_functionalized_v2 (because then we're really signing up for BC on how we serialize the view reconstruction info)

zou3519 replied:
Oh, but we do have the view information at this point in time, so we shouldn't need to guess it. @bdhirsh I agree with your point, please ignore the last comment lol

laithsakka replied:
if you mean we do not need to guess if the view came from an alias, i guess that is true we can probably also call check the recorded views. 

But  not sure what do you mean by " why we need the as_strided "

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1098,6 +1098,16 @@ def parallel_and(*args: BoolLikeType) -> BoolLikeType:
+    """
+    if any(statically_known_true(torch.sym_not(a)) for a in args):
+        return False
+    return all(args)
```
By in large, we are not using `parallel_and` or `parallel_or` (because we have mostly solved these problems in other ways, e.g., 0/1 specialization), so I'm not that hot on adding this new variation on the function. Actually honestly we should probably remove the old ones.

laithsakka replied:
I see I can just use statically_known_true directly there and remove this abstraction.

laithsakka replied:
https://github.com/pytorch/pytorch/pull/138135

laithsakka replied:
I got red of it , but now looking at its really not probably what i originally intended. 
I wanted all(statically_known_true(a)) for a in args) anyhow i inlined it up there.

laithsakka committed:
Update on "Rewrite the check for alias simplification in auto_functionalize_v2 in a way that guarantee no guards are created"


The alias simplification is best effort, its better not to introduce guards than to do the alias simplification. 
if the view is originally coming from an alias operation, MOST likely we should be able to statically know that, if Not
Well its ok then not to use alias. 

this way we guarantee that there are no guards added. 


[ghstack-poisoned]

laithsakka commented:
remove parallel_and_statically_known_true and rewrite using statically_known_true

laithsakka committed:
Update on "Rewrite the check for alias simplification in auto_functionalize_v2 in a way that guarantee no guards are created"


The alias simplification is best effort, its better not to introduce guards than to do the alias simplification. 
if the view is originally coming from an alias operation, MOST likely we should be able to statically know that, if Not
Well its ok then not to use alias. 

this way we guarantee that there are no guards added. 


[ghstack-poisoned]
